include/common/aarch64/console_macros.S

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2017, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#ifndef __CONSOLE_MACROS_S__
+#define __CONSOLE_MACROS_S__
+
+#include <asm_macros.S>
+#include <console.h>
+
+/*
+ * This file defines macros for use by console drivers that want to hook into
+ * the generic console API. Each console driver needs to include this file, add
+ * a declare_console directive and implement at least func_console_init. All
+ * other functions are optional. Functions need to be declared through these
+ * macros to provide backwards-compatiblity with the previous console API for
+ * older platform code. All functions need to take care to obey the register
+ * restrictions outlined below and not change any register not explicitly listed
+ * in the clobber list. In addition to the listed registers, implementations are
+ * always allowed to use the intra-procedure-call registers (X16 and X17) as
+ * temporary scratch registers within a single function.
+ */
+
+	/* ------------------------------------------------
+	 * int console_<driver>_init(uintptr_t baseaddr,
+	 *                           ...args...)
+	 * Initializes the console. Will be called directly
+	 * from console_<driver>_register and receive any
+	 * arguments in x0-x7 unchanged from it. x0 is
+	 * intended to be a base address pointer, but can
+	 * be used for any purpose. It will be stored in
+	 * the console structure and passed in to every
+	 * other callback for this console.
+	 * In: x0 - driver-specific base address
+	 *     x1...x7 - optional extra arguments
+	 * Out: x0 - 1 for success, 0 for error
+	 * Clobber list: x0 - x7
+	 * ------------------------------------------------
+	 */
+.macro func_console_init _driver
+	func console_\_driver\()_init
+	.globl console_core_init
+	.weak console_core_init
```

Comment:
You're right that .globl doesn't mean strong (sorry about that), although I think declaring a symbol as weak effectively makes it global. I don't think allowing multiple weak definitions is very robust, due to the ambiguity as to which one the linker chooses. We intend to ban this pattern in future to comply as best we can with certain MISRA rules. We should only allow at most one weak and one strong definition.

If these console_core_xxx declarations were hidden behind a build flag (e.g. ALLOW_MULTIPLE_CONSOLES) then there should only be a single (strong) definition of console_core_xxx for unmigrated platforms, so declaring them as weak would not be necessary. Migrated platforms can set ALLOW_MULTIPLE_CONSOLES to 1 so that these functions are not declared. ALLOW_MULTIPLE_CONSOLES could default to the value of ERROR_DEPRECATED.